addons/web/static/src/legacy/js/public/public_widget.js

```diff
@@ -143,78 +153,236 @@ var PublicWidget = Widget.extend({
      * @param {Object} [options]
      */
     init: function (parent, options) {
-        this._super.apply(this, arguments);
+        mixins.PropertiesMixin.init.call(this);
+        this.setParent(parent);
         this.options = options || {};
     },
+    /**
+     * Method called between @see init and @see start. Performs asynchronous
+     * calls required by the rendering and the start method.
+     *
+     * This method should return a Promose which is resolved when start can be
+     * executed.
+     *
+     * @returns {Promise}
+     */
+    willStart: function () {
+        var proms = [];
+        if (this.jsLibs || this.cssLibs || this.assetLibs) {
+            proms.push(loadBundle(this));
+        }
+        return Promise.all(proms);
+    },
+    /**
+     * Method called after rendering. Mostly used to bind actions, perform
+     * asynchronous calls, etc...
+     *
+     * By convention, this method should return an object that can be passed to
+     * Promise.resolve() to inform the caller when this widget has been initialized.
+     *
+     * Note that, for historic reasons, many widgets still do work in the start
+     * method that would be more suited to the willStart method.
+     *
+     * @returns {Promise}
+     */
+    start: function () {
+        return Promise.resolve();
+    },
     /**
      * Destroys the widget and basically restores the target to the state it
      * was before the start method was called (unlike standard widget, the
      * associated $el DOM is not removed, if this was instantiated thanks to the
      * selector property).
      */
     destroy: function () {
-        if (this.selector) {
-            var $oldel = this.$el;
-            // The difference with the default behavior is that we unset the
-            // associated element first so that:
-            // 1) its events are unbinded
```

Comment:
Did not you lost that part? :grimacing: Strange, because if you did, I guess there should be many bugs like "animations still running during edit mode while they should not" etc.

Reply:
I thought that the call to this would be enough: https://github.com/odoo/odoo/blob/60a72cfd3c812ad9263e240b8ce343357b8234b7/addons/web/static/src/legacy/js/core/mixins.js#L344

But I did not dive deeper into that code, there might be reasons I'm unaware of that some events would not be unbound. I would not mind calling `this._undelegateEvents` just to be safe, up to you!

Reply:
@detrouxdev Why would you not want `this.$el.off('.widget_events');` made by `this._undelegateEvents` to be done? It is not done in the destroy of the base Widget class... but because the whole `$el` is removed from the DOM (thus indirectly binding the events on them), which is not the case for public widgets with a `selector`. So in that case, you would have an element on the DOM, whose public widget is destroyed but whose events are still bound.

> I would not mind calling it just to be safe, up to you!

It's not up to me, it's up to you :wink: Could you try by testing it to see if I am right? Create a public widget readonly-mode-only (not instantiated in edit mode) targeting buttons, set up a click event... and check if currently the click handler is still there when entering edit mode, for example :)

And if I am wrong, tell me why :wink: But I don't think this related to the EventDispatcherMixin's destroy you linked, that one should only handle custom event, not native events :wink:

Reply:
My assumptions were wrong, indeed the event dispatcher only handles custom events, which is painfully obvious now that I have re-read the code. I also did the test just to be sure, and you were right.

I have adapted the code to call `this._undelegateEvents` on destroy.